### PR TITLE
Feature: expand, plus broadcasting for +/-/*/divide

### PR DESCRIPTION
**Pull request by dsyme**
Implement broadcasting

* [x] changes `tensor.Extend(...)` into a full `tensor.Expand(...)` that broadcasts through prefix of dimensions plus expanding `1` --> `N`.

* [x] implement broadcasting for `+`, `-`, `*`, `/`

* [x] add testing for `Expand`

* [x] add testing for broadcasting for `+`, `-`, `*`, `/`

* [x] add testing for derivatives of broadcasting for `+`, `-`, `*`, `/`

I did additional tested by commenting out the existing Scalar/Vector/Matrix extension.  Will discuss with @gbaydin on Friday

As an aside, to match PyTorch the `Expand` should use a view with a stride=0 in the appropriate dimensions (and thus not reallocate anything), but this requires strided views.  I'll add a TBD about that in #69

### CODE REVIEW

**Comment by gbaydin** [comment]
Thank you! The derivative tests are a bit hard to follow. Can we please improve those. I will look through the code and check everything carefully as well.

**Comment by dsyme** [comment]
I've added the explicit tests for the derivatives of `expand` here https://github.com/DiffSharp/DiffSharp/pull/85/files#diff-79888ca94c742a41ec58bcc44dea5c97R73

I think this is now ready.  Would appreciate python code to test the forward derivatives

**Comment by gbaydin** [comment]
Thank you. I reviewed these changes and I'm convinced they are fine pending some changes in testing that I will implement.
